File: UVa/10639 - Square Puzzle/Square Puzzle.java
import java.io.BufferedOutputStream;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.Scanner;
import java.util.function.Predicate;
import java.io.BufferedInputStream;

public class Main {
	public static void main(String[] args) throws IOException {
		Scanner scanner;
		
		PrintWriter printWriter = new PrintWriter(new BufferedOutputStream(System.out));
		
		if (DEBUG) {
			scanner = new Scanner(new BufferedInputStream(new FileInputStream("f:\\OJ\\uva_in.txt")));
		} else {
			scanner = new Scanner(System.in);
		}
		
		int p = scanner.nextInt();
		for (int i = 0; i < p; ++i) {
			int n, m;
			n = scanner.nextInt();
			m = scanner.nextInt();
			
			Main solution = new Main();
			solution.n = n;
			solution.m = m;
			for (int j = 0; j < n; ++j) {
				Polygon polygon = new Polygon();
				int k = scanner.nextInt();
				polygon.pointNum = k;
				for (int a = 0; a < k; ++a) {
					int x = scanner.nextInt();
					int y = scanner.nextInt();
					Point point = new Point(x, y);
					polygon.points.add(point);
				}
				solution.polygons.add(polygon);
			}
			
			boolean ans = solution.solve(); 
			if (ans) { 
				  printWriter.println("yes"); 
			} else { 
				printWriter.println("no"); 
			}
			printWriter.flush();
		}
		
		scanner.close();
	}
	
	public boolean solve() {
		filled = new int[m][m];
		
		List<List<int[][]>> puzzles = initList(n);
		for (int i = 0; i < n; ++i) {
			int[][] fill = scanLineFill(polygons.get(i));
			
			initFeasibleFill(puzzles.get(i), fill);
		}
		
		//printPuzzles(puzzles);
		boolean[] visited = new boolean[n];
		boolean ok = dfs(0, puzzles, filled, visited);
		return ok;
	}
	
	private int[][] scanLineFill(Polygon polygon) {
		int ymin = polygon.points.stream().map(Point::getY).min(Integer::compare).get();
		int ymax = polygon.points.stream().map(Point::getY).max(Integer::compare).get();
		int xmin = polygon.points.stream().map(Point::getX).min(Integer::compare).get();
		int xmax = polygon.points.stream().map(Point::getX).max(Integer::compare).get();
		
		//System.out.println("ymax:"+ ymax + " ymin:" + ymin);
		List<Edge>[] netEdges;
		List[] originLists = new List[ymax - ymin + 1];
		for (int y = ymin; y <= ymax; ++y) {
			originLists[y - ymin] = new ArrayList<>(); 
		}
		netEdges = (List<Edge>[])originLists;
		initScanLineNewEdgeTable(netEdges, polygon, ymin, ymax);
		//printNewEdgeTable(netEdges);
		int[][] fill = new int[ymax - ymin][xmax - xmin];
		horizonEdgeFill(polygon, fill, xmin, ymin, ymax);
		calFill(netEdges, ymin, ymax, xmin, xmax, fill);
		return fill;
	}
	
	private int[][] calFill(List<Edge>[] edgeLists, int ymin, int ymax, int xmin, int xmax, int[][] fill) {
		List<Edge> aetEdges = new ArrayList<>(); 
		
		for (int y = ymin; y <= ymax; ++y) {
			insertNetListToAet(edgeLists[y - ymin], aetEdges);
			
			fillAetScanLine(aetEdges, fill, xmin, xmax, y, ymin, ymax);
			
			removeNonActiveEdgeFromAet(aetEdges, y);
			updateAndResortAet(aetEdges);
		}
		
		return fill;
	}
	
	private void horizonEdgeFill(Polygon polygon, int[][] fill, int xmin, int ymin, int ymax) {
		for (int i = 0; i < polygon.pointNum; ++i) {
			Point startPoint = polygon.points.get(i);
			Point endPoint = polygon.points.get((i + 1) % polygon.pointNum);
			if (startPoint.y != endPoint.y) {
				continue;
			}
			
			int cury = startPoint.y;
			if (cury == ymax) {
				break;
			}
			int startx = Math.min(startPoint.x, endPoint.x);
			int endx = Math.max(startPoint.x, endPoint.x);
			for (int j = startx; j < endx; ++j) {
				fill[cury - ymin][j - xmin] = FillType.FULL_FILL.getCode();
			}
		}
	}
	
	private void updateAndResortAet(List<Edge> edges) {
		edges.stream().forEach(edge -> {
			edge.x += edge.dx;
		});
		
		edges.sort(Comparator.comparing(Edge::getx));
	}
	
	private void removeNonActiveEdgeFromAet(List<Edge> aet, int y) {
		aet.removeIf(new Predicate<Edge>() {

			@Override
			public boolean test(Edge t) {
				return t.ymax == y;
			}
		});
	}

	private void fillAetScanLine(List<Edge> aetEdges, int[][] fill, int xmin, int xmax, int y, int ymin, int ymax) {
		int size = aetEdges.size();
		for (int j = 0; j < size - 1; j += 2) {
			fillTwoPoint(aetEdges.get(j), aetEdges.get(j + 1), fill, xmin, xmax, y, ymin, ymax);
		}
	}
	
	private void fillTwoPoint(Edge a, Edge b, int[][] fill, int xmin, int xmax, int y, int ymin, int ymax) {
		if (y == ymax) {
			return;
		}
		int xstart = a.x;
		int xend = b.x;
		
		if (a.dx == 1) {
			fill[y - ymin][xstart - xmin] = FillType.DIAG_RIGHT_DOWN.getCode();
			++xstart;
		} else if (a.dx == -1) {
			fill[y - ymin][xstart - 1 - xmin] = FillType.DIAG_RIGHT_UP.getCode();
		}
		
		if (b.dx == -1) {
			fill[y - ymin][xend - 1 - xmin] = FillType.DIAG_LEFT_DOWN.getCode();
			--xend;
		} else if (b.dx == 1) {
			fill[y - ymin][xend - xmin] = FillType.DIAG_LEFT_UP.getCode();
		}
		
		for (int x = xstart; x < xend; ++x) {
			fill[y - ymin][x - xmin] = FillType.FULL_FILL.getCode();
		}
	}
	
	private void insertNetListToAet(List<Edge> net, List<Edge> aet) {
		if (net.isEmpty()) {
			return;
		}
		
		aet.addAll(net);
		
		aet.sort(Comparator.comparing(Edge::getx));
	}
	
	private void initScanLineNewEdgeTable(List<Edge>[] lists, Polygon polygon, int ymin, int ymax) {
		for (int i = 0; i < polygon.pointNum; ++i) {
			Point startPoint = polygon.points.get(i);
			Point endPoint = polygon.points.get((i + 1) % polygon.pointNum);
			Point sstartPoint = polygon.points.get((i - 1 + polygon.pointNum) % polygon.pointNum);
			Point eendPoint = polygon.points.get((i + 2) % polygon.pointNum);
			
			if (startPoint.y == endPoint.y) {
				continue;
			}
			
			Edge edge = new Edge();
			edge.dx = (endPoint.x - startPoint.x) / (endPoint.y - startPoint.y);
			if (endPoint.y > startPoint.y) {
				edge.x = startPoint.x;
				if (eendPoint.y >= endPoint.y) {
					edge.ymax = endPoint.y - 1;
				} else {
					edge.ymax = endPoint.y;
				}
				
				lists[startPoint.y - ymin].add(0, edge);
			} else {
				edge.x = endPoint.x;
				if (sstartPoint.y >= startPoint.y) {
					edge.ymax = startPoint.y - 1;
				} else {
					edge.ymax = startPoint.y;
				}
				
				lists[endPoint.y - ymin].add(0, edge);
			}
		}
	}
	
	private void printNewEdgeTable(List<Edge>[] netEdges) {
		System.out.println("length:" + netEdges.length);
		for (int i = 0; i < netEdges.length; ++i) {
			if (netEdges[i].size() == 0) {
				System.out.println("empty");
				continue;
			}
			System.out.println("----------------");
			for (int j = 0; j < netEdges[i].size(); ++j) {
				System.out.println("x:" + netEdges[i].get(j).x + " dx:" + netEdges[i].get(j).dx + " ymax:" + netEdges[i].get(j).ymax);
			}
			System.out.println("----------------");
		}
	}
	
	private void printPuzzles(List<List<int[][]>> puzzles) {
		System.out.println("puzzles.size():" + puzzles.size());
		for (int i = 0; i < puzzles.size(); ++i) {
			List<int[][]> puzzle = puzzles.get(i);
			System.out.println("puzzle.size():" + puzzle.size());
			System.out.println("---------------");
			for (int j = 0; j < puzzle.size(); ++j) {
				System.out.println(Arrays.deepToString(puzzle.get(j)));
			}
			System.out.println("---------------");
		}
	}
	
	private int[][] rotate(int[][] fill) {
		int row = fill.length;
		int col = fill[0].length;
		
		int[][] result = new int[col][row];
		
		for (int i = 0; i < row; ++i) {
			for (int j = 0; j < col; ++j) {
				if (fill[i][j] == FillType.NOT_FILL.getCode() || fill[i][j] == FillType.FULL_FILL.getCode()) {
					result[col - 1 - j][i] = fill[i][j];
				} else if (fill[i][j] == FillType.DIAG_LEFT_UP.getCode()) {
					result[col - 1 - j][i] = FillType.DIAG_RIGHT_UP.getCode();
				} else if (fill[i][j] == FillType.DIAG_RIGHT_UP.getCode()) {
					result[col - 1 - j][i] = FillType.DIAG_RIGHT_DOWN.getCode();
				} else if (fill[i][j] == FillType.DIAG_RIGHT_DOWN.getCode()) {
					result[col - 1 - j][i] = FillType.DIAG_LEFT_DOWN.getCode();
				} else if (fill[i][j] == FillType.DIAG_LEFT_DOWN.getCode()) {
					result[col - 1 - j][i] = FillType.DIAG_LEFT_UP.getCode();
				}
			}
		}
		return result;
	}
	
	private boolean fillEqual(int[][] source, int[][] target) {
		int sourceRow = source.length;
		int sourceCol = source[0].length;
		
		int targetRow = target.length;
		int targetCol = target[0].length;
		
		if (sourceRow != targetRow) {
			return false;
		}
		
		if (sourceCol != targetCol) {
			return false;
		}
		
		for (int i = 0; i < sourceRow; ++i) {
			for (int j = 0; j < sourceCol; ++j) {
				if (source[i][j] != target[i][j]) {
					return false;
				}
			}
		}
		
		return true;
	}
	
	private List<List<int[][]>> initList(int n) {
		List<List<int[][]>> puzzles = new ArrayList<>();
		for (int i = 0; i < n; ++i) {
			puzzles.add(new ArrayList<>());
		}
		
		return puzzles;
	}
	
	private void initFeasibleFill(List<int[][]> list, int[][] fill) {
		list.add(fill);
		int[][] result = fill;
		for (int i = 0; i < 3; ++i) {
			result = rotate(result);
			boolean exist = false;
			for (int j = 0; j < list.size(); ++j) {
				if (fillEqual(list.get(j), result)) {
					exist = true;
					break;
				}
			}
			
			if (!exist) {
				list.add(result);
			}
		}
	}
	
	boolean dfs(int cur, List<List<int[][]>> fills, int[][] filled, boolean[] visited) {
		if (cur == n) {
			boolean ok = true;
			for (int i = 0; i < m; ++i) {
				for (int j = 0; j < m; ++j) {
					if (filled[i][j] != FillType.FULL_FILL.getCode()) {
						ok = false;
						break;
					}
				}
			}
			
			return ok;
		}
		
		int startx = 0, starty = 0;
		boolean found = false;
		for (int i = 0; i < m && !found; ++i) {
			for (int j = 0; j < m && !found; ++j) {
				if (filled[i][j] != FillType.FULL_FILL.getCode()) {
					found = true;
					startx = i;
					starty = j;
					break;
				}
			}
		}
		
		for (int i = 0; i < n; ++i) {
			if (!visited[i]) {
				visited[i] = true;
				List<int[][]> curFeasibleFill = fills.get(i);
				
				for (int k = 0; k < curFeasibleFill.size(); ++k) {
					int[][] curFill = curFeasibleFill.get(k);
					int row = curFill.length;
					int col = curFill[0].length;
					if (startx + row > m) continue;
					if (starty + col > m) continue;
					
					
					fillCurSquare(curFill, startx, starty, filled);
					boolean ok = checkCurFillSquare(curFill, startx, starty, filled);
					if (!ok) {
						unFillCurSquare(curFill, startx, starty, filled);
						continue;
					}
					
					if (dfs(cur + 1, fills, filled, visited)) {
						return true;
					} 
					unFillCurSquare(curFill, startx, starty, filled);
						
				}
				
				visited[i] = false;
			}
		}
		List<int[][]> curFeasibleFill = fills.get(cur);
		for (int i = 0; i < m; ++i) {
			for (int j = 0; j < m; ++j) {
				if (filled[i][j] != FillType.FULL_FILL.getCode()) {
					for (int k = 0; k < curFeasibleFill.size(); ++k) {
						int[][] curFill = curFeasibleFill.get(k);
						int row = curFill.length;
						int col = curFill[0].length;
						if (i + row > m) continue;
						if (j + col > m) continue;
						
						for (int a = 0; a <= m - row; ++a) {
							for (int b = 0; b <= m - col; ++b) {
								fillCurSquare(curFill, a, b, filled);
								boolean ok = checkCurFillSquare(curFill, a, b, filled);
								if (!ok) {
									unFillCurSquare(curFill, a, b, filled);
									continue;
								}
								
								if (dfs(cur + 1, fills, filled, visited)) {
									return true;
								} 
								unFillCurSquare(curFill, a, b, filled);
							}
						}
					}
				}
			}
		}
		
		return false;
	}
	
	private void fillCurSquare(int[][] square, int xoffset, int yoffset, int[][] filled) {
		int row = square.length;
		int col = square[0].length;
		
		for (int i = 0; i < row; ++i) {
			for (int j = 0; j < col; ++j) {
				filled[xoffset + i][yoffset + j] += square[i][j];
			}
		}
	}
	
	private void unFillCurSquare(int[][] square, int xoffset, int yoffset, int[][] filled) {
		int row = square.length;
		int col = square[0].length;
		
		for (int i = 0; i < row; ++i) {
			for (int j = 0; j < col; ++j) {
				filled[xoffset + i][yoffset + j] -= square[i][j];
			}
		}
	}
	
	private boolean checkCurFillSquare(int[][] square, int xoffset, int yoffset, int[][] filled) {
		int row = square.length;
		int col = square[0].length;
		
		for (int i = 0; i < row; ++i) {
			for (int j = 0; j < col; ++j) {
				if (filled[i][j] != FillType.NOT_FILL.getCode() && 
						filled[i][j] != FillType.FULL_FILL.getCode() &&
						filled[i][j] != FillType.DIAG_LEFT_UP.getCode() &&
						filled[i][j] != FillType.DIAG_RIGHT_DOWN.getCode() &&
						filled[i][j] != FillType.DIAG_LEFT_DOWN.getCode() &&
						filled[i][j] != FillType.DIAG_RIGHT_UP.getCode()) {
					return false;
				}
			}
		}
		
		return true;
	}
	
	
	private int[][] filled;
	private static final boolean DEBUG = false;
	
	private int m, n;
	private List<Polygon> polygons = new ArrayList<>();
	
	static class Point {
		int x, y;
		
		Point(int x, int y) {
			this.x = x;
			this.y = y;
		}
		
		int getX() {
			return x;
		}
		
		int getY() {
			return y;
		}
		
	}
	
	static class Polygon {
		List<Point> points = new ArrayList<>();
		int pointNum;
	}
	
	static class Edge {
		int x;
		int dx;
		int ymax;
		
		int getx() {
			return x;
		}
	}
	
	enum FillType {
		NOT_FILL(0),
		FULL_FILL(1),
		DIAG_LEFT_UP(-1),
		DIAG_RIGHT_DOWN(2),
		DIAG_LEFT_DOWN(-3),
		DIAG_RIGHT_UP(4);
		
		private int code;
		
		private FillType(int code) {
			this.code = code;
		}
		
		public int getCode() {
			return code;
		}
	}

}

File: UVa/10639 - Square Puzzle/Square Puzzle.cpp
#include <iostream>
#include <fstream>
#include <algorithm>
#include <vector>
#include <cstring>
#include <functional>

using namespace std;

const int MAXM = 7;
const int NOT_FILL = 0, FULL_FILL = 1, DIAG_LEFT_UP = -1, DIAG_RIGHT_DOWN = 2, DIAG_LEFT_DOWN = -3, DIAG_RIGHT_UP = 4;

struct Point
{
    int x, y;
};

struct Polygon
{
    vector<Point> points;
};

struct Edge
{
    int x, dx, ymax;

    bool operator < (const Edge& other) const
    {
        return x < other.x;
    }
};

struct FillSquare
{
    int fill[MAXM][MAXM];
    int row, col;
};

int n, m;
vector<Polygon> polygons;
FillSquare filled;
vector<vector<FillSquare>> puzzles;
int xmin, xmax, ymin, ymax;

void fastio()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);
}

void printNewEdgeTable(vector<vector<Edge>>& netEdges) 
{
    cout << "length:" << netEdges.size() << endl;
    for (int i = 0; i < netEdges.size(); ++i) {
        if (netEdges[i].size() == 0) {
            cout << "empty" << endl;
            continue;
        }
        cout << "----------------" << endl;
        for (int j = 0; j < netEdges[i].size(); ++j) {
            cout << "x:" << netEdges[i][j].x  << " dx:"  << netEdges[i][j].dx  << " ymax:" << netEdges[i][j].ymax << endl;
        }
        cout << "----------------" << endl;
    }
}

void printPuzzles(vector<vector<FillSquare>> puzzles) 
{
    cout << "puzzles.size():" << puzzles.size() << endl;
    for (int i = 0; i < puzzles.size(); ++i) {
        vector<FillSquare>& puzzle = puzzles[i];
        cout << "puzzle.size():"  << puzzle.size() << endl;
        cout <<"---------------" << endl;
        for (int j = 0; j < puzzle.size(); ++j) {
            FillSquare& fillSquare = puzzle[j];
            for (int a = 0; a < fillSquare.row; ++a) {
                for (int b = 0; b < fillSquare.col; ++b) {
                    cout << fillSquare.fill[a][b] << " ";
                }
                cout << endl;
            }
        }
        cout <<"---------------" << endl;
    }
}
void calXYMinMax(Polygon& polygon, int& xmin, int& xmax, int& ymin, int& ymax)
{
    xmin = ymin = INT_MAX;
    xmax = ymax = INT_MIN;

    for (int i = 0; i < polygon.points.size(); ++i) {
        const Point& point = polygon.points[i];

        xmin = min(xmin, point.x);
        xmax = max(xmax, point.x);

        ymin = min(ymin, point.y);
        ymax = max(ymax, point.y);
    }
}



void initScanLineNewEdgeTable(vector<vector<Edge>>& edges, const Polygon& polygon)
{
    int pointNum = polygon.points.size();
    for (int i = 0; i < pointNum; ++i) {
        const Point& startPoint = polygon.points[i];
        const Point& endPoint = polygon.points[(i + 1) % pointNum];
        const Point& sstartPoint = polygon.points[(i - 1 + pointNum) % pointNum];
        const Point& eendPoint = polygon.points[(i + 2) % pointNum];

        if (startPoint.y == endPoint.y) {
            continue;
        }

        Edge edge;
        edge.dx = (endPoint.x - startPoint.x) / (endPoint.y - startPoint.y);
        if (endPoint.y > startPoint.y) {
            edge.x = startPoint.x;
            if (eendPoint.y >= endPoint.y) {
                edge.ymax = endPoint.y - 1;
            } else {
                edge.ymax = endPoint.y;
            }

            edges[startPoint.y - ymin].insert(edges[startPoint.y - ymin].begin(), edge);
        } else {
            edge.x = endPoint.x;
            if (sstartPoint.y >= startPoint.y) {
                edge.ymax = startPoint.y - 1;
            } else {
                edge.ymax = startPoint.y;
            }

            edges[endPoint.y - ymin].insert(edges[endPoint.y - ymin].begin(), edge);
        }
    }
}

void insertNetListToAet(vector<Edge>& net, vector<Edge>& aet)
{
    if (net.size() == 0) {
        return;
    }

    for (auto& e : net) {
        aet.push_back(e);
    }

    sort(aet.begin(), aet.end());
}

void fillTwoPoint(Edge& a, Edge& b, FillSquare& fillSquare, int y)
{
    if (y == ymax) {
        return;
    }
    int xstart = a.x;
    int xend = b.x;

    if (a.dx == 1) {
        fillSquare.fill[y - ymin][xstart - xmin] = DIAG_RIGHT_DOWN;
        ++xstart;
    } else if (a.dx == -1) {
        fillSquare.fill[y - ymin][xstart - 1 - xmin] = DIAG_RIGHT_UP;
    }

    if (b.dx == -1) {
        fillSquare.fill[y - ymin][xend - 1 - xmin] = DIAG_LEFT_DOWN;
        --xend;
    } else if (b.dx == 1) {
        fillSquare.fill[y - ymin][xend - xmin] = DIAG_LEFT_UP;
    }

    for (int x = xstart; x < xend; ++x) {
        fillSquare.fill[y - ymin][x - xmin] = FULL_FILL;
    }
}

void fillAetScanLine(vector<Edge>& aetEdges, FillSquare& fillSquare, int y)
{
    int size = aetEdges.size();
    for (int i = 0; i < size - 1; i += 2) {
        fillTwoPoint(aetEdges[i], aetEdges[i + 1], fillSquare, y);
    }
}

bool isEdgeOutOfActive(Edge e, int y)
{
    return e.ymax == y;
}

void removeNonActiveEdgeFromAet(vector<Edge>& aet, int y)
{
    aet.erase(remove_if(aet.begin(), aet.end(), bind2nd(ptr_fun(isEdgeOutOfActive), y)));
}

void updateAetEdgeInfo(Edge& e)
{
    e.x += e.dx;
}

void updateAndResortAet(vector<Edge>& aet)
{
    for_each(aet.begin(), aet.end(), updateAetEdgeInfo);
    sort(aet.begin(), aet.end());
}

void calFill(vector<vector<Edge>>& edges, FillSquare& fillSquare)
{
    vector<Edge> activeEdges;
    for (int y = ymin; y <= ymax; ++y) {
        insertNetListToAet(edges[y - ymin], activeEdges);

        fillAetScanLine(activeEdges, fillSquare, y);

        removeNonActiveEdgeFromAet(activeEdges, y);
        updateAndResortAet(activeEdges);
    }
}

void scanLineFill(Polygon& polygon, FillSquare& fillSquare)
{
    int xmin, xmax, ymin, ymax;
    calXYMinMax(polygon, xmin, xmax, ymin, ymax);
    vector<vector<Edge>> edges(ymax - ymin + 1);

    initScanLineNewEdgeTable(edges, polygon);
    printNewEdgeTable(edges);
    fillSquare.row = ymax - ymin;
    fillSquare.col = xmax - xmin;
    calFill(edges, fillSquare);
}

FillSquare rotate(FillSquare& source)
{
    int row = source.row;
    int col = source.col;

    FillSquare result;
    result.row = col;
    result.col = row;

    for (int i = 0; i < row; ++i) {
        for (int j = 0; j < col; ++j) {
            if (source.fill[i][j] == NOT_FILL || source.fill[i][j] == FULL_FILL) {
                result.fill[col - 1 - j][i] = source.fill[i][j];
            } else if (source.fill[i][j] == DIAG_LEFT_UP) {
                result.fill[col - 1 - j][i] = DIAG_RIGHT_UP;
            } else if (source.fill[i][j] == DIAG_RIGHT_UP) {
                result.fill[col - 1 - j][i] = DIAG_RIGHT_DOWN;
            } else if (source.fill[i][j] == DIAG_RIGHT_DOWN) {
                result.fill[col - 1 - j][i] = DIAG_LEFT_DOWN;
            } else if (source.fill[i][j] == DIAG_LEFT_DOWN) {
                result.fill[col - 1 - j][i] = DIAG_LEFT_UP;
            }
        }
    }

    return result;
}

bool fillEqual(FillSquare& source, FillSquare& target)
{
    int sourceRow = source.row;
    int sourceCol = source.col;

    int targetRow = target.row;
    int targetCol = target.col;

    if (sourceRow != targetRow) {
        return false;
    }

    if (sourceCol != targetCol) {
        return false;
    }

    for (int i = 0; i < sourceRow; ++i) {
        for (int j = 0; j < sourceCol; ++j) {
            if (source.fill[i][j] != target.fill[i][j]) {
                return false;
            }
        }
    }

    return true;
}

void initFeasibleFill(vector<FillSquare>& list, FillSquare& fill)
{
    list.push_back(fill);
    FillSquare result = fill;
    for (int i = 0; i < 3; ++i) {
        result = rotate(result);
        bool exist = false;
        for (int j = 0; j < list.size(); ++j) {
            if (fillEqual(list[j], result)) {
                exist = true;
                break;
            }
        }

        if (!exist) {
            list.push_back(result);
        }
    }
}

void unFillCurSquare(FillSquare& square, int xoffset, int yoffset, FillSquare& filled)
{
    int row = square.row;
    int col = square.col;

    for (int i = 0; i < row; ++i) {
        for (int j = 0; j < col; ++j) {
            filled.fill[xoffset + i][yoffset + j] -= square.fill[i][j];
        }
    }
}

void fillCurSquare(FillSquare& square, int xoffset, int yoffset, FillSquare& filled)
{
    int row = square.row;
    int col = square.col;

    for (int i = 0; i < row; ++i) {
        for (int j = 0; j < col; ++j) {
            filled.fill[xoffset + i][yoffset + j] += square.fill[i][j];
        }
    }
}

bool checkCurFillSquare(FillSquare& square, int xoffset, int yoffset, FillSquare& filled)
{
    int row = square.row;
    int col = square.col;

    for (int i = 0; i < row; ++i) {
        for (int j = 0; j < col; ++j) {
            if (filled.fill[i][j] != NOT_FILL &&
                    filled.fill[i][j] != FULL_FILL &&
                    filled.fill[i][j] != DIAG_LEFT_UP &&
                    filled.fill[i][j] != DIAG_RIGHT_DOWN &&
                    filled.fill[i][j] != DIAG_LEFT_DOWN &&
                    filled.fill[i][j] != DIAG_RIGHT_UP) {
                return false;
            }
        }
    }

    return true;
}

bool dfs(int cur, vector<vector<FillSquare>>& fills, FillSquare& filled)
{
    if (cur == n) {
        bool ok = true;
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < m; ++j) {
                if (filled.fill[i][j] != FULL_FILL) {
                    ok = false;
                    break;
                }
            }
        }

        return ok;
    }

    vector<FillSquare>& curFeasibleFill = fills[cur];
    for (int i = 0; i < m; ++i) {
        for (int j = 0; j < m; ++j) {
            if (filled.fill[i][j] != FULL_FILL) {
                for (int k = 0; k < curFeasibleFill.size(); ++k) {
                    FillSquare& curFill = curFeasibleFill[k];
                    int row = curFill.row;
                    int col = curFill.col;
                    if (i + row > m) continue;
                    if (j + col > m) continue;

                    for (int a = 0; a <= m - row; ++a) {
                        for (int b = 0; b <= m - col; ++b) {
                            fillCurSquare(curFill, a, b, filled);
                            bool ok = checkCurFillSquare(curFill, a, b, filled);
                            if (!ok) {
                                unFillCurSquare(curFill, a, b, filled);
                                continue;
                            }

                            if (dfs(cur + 1, fills, filled)) {
                                return true;
                            }
                            unFillCurSquare(curFill, a, b, filled);
                        }
                    }
                }
            }
        }
    }


    for (int i = 0; i < curFeasibleFill.size(); ++i) {
        FillSquare& curFill = curFeasibleFill[i];
        int row = curFill.row;
        int col = curFill.col;

        for (int j = 0; j <= m - row; ++j) {
            for (int k = 0; k <= m - col; ++k) {
                fillCurSquare(curFill, j, k, filled);
                bool ok = checkCurFillSquare(curFill, j, k, filled);
                if (!ok) {
                    unFillCurSquare(curFill, j, k, filled);
                    continue;
                }

                if (dfs(cur + 1, fills, filled)) {
                    return true;
                }
                unFillCurSquare(curFill, j, k, filled);
            }
        }
    }

    return false;
}

bool solve()
{
    memset(&filled, 0x00, sizeof filled);

    for (int i = 0; i < n; ++i) {
        FillSquare fillSquare;
        memset(&fillSquare, 0x00, sizeof fillSquare);
        scanLineFill(polygons[i], fillSquare);
        initFeasibleFill(puzzles[i], fillSquare);
    }
    
    printPuzzles(puzzles);
    bool ok = dfs(0, puzzles, filled);
    return ok;
}

int main()
{
    fastio();
    #ifndef ONLINE_JUDGE
        ifstream fin("f:\\OJ\\uva_in.txt");
        streambuf* cinback = cin.rdbuf(fin.rdbuf());
    #endif // ONLINE_JUDGE

    int p;
    cin >> p;
    while (p--) {
        cin >> n >> m;
        polygons.clear();
        for (int i = 0; i < n; ++i) {
            int k;
            cin >> k;
            Polygon polygon;
            for (int j = 0; j < k; ++j) {
                int x, y;
                cin >> x >> y;
                polygon.points.push_back((Point){x, y});
            }
            polygons.push_back(polygon);
        }

        bool ans = solve();
        if (ans) {
            cout << "yes";
        } else {
            cout << "no";
        }
    }

    #ifndef ONLINE_JUDGE
        cin.rdbuf(cinback);
    #endif // ONLINE_JUDGE
    return 0;
}
